src/richie/apps/courses/models/role.py

```diff
@@ -0,0 +1,91 @@
+"""
+Declare and configure the models for the courses application
+"""
+from django.contrib.auth.models import Group
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+
+from cms.api import Page
+
+from ...core.helpers import get_permissions
+from ..defaults import ADMIN, EDITOR, PAGE_ROLES
+
+
+class PageRole(models.Model):
+    """A model to define and control by roles the permissions related to a page."""
+
+    ROLE_CHOICES = ((ADMIN, _("Admin")), (EDITOR, _("Editor")))
+
+    role = models.CharField(
+        choices=ROLE_CHOICES,
+        max_length=20,
+        verbose_name=_("role"),
+        help_text=_(
+            "A role describes all the permissions that should be "
+            "granted to the user group."
+        ),
+    )
+    page = models.ForeignKey(
+        to=Page,
+        related_name="roles",
+        verbose_name=_("page"),
+        help_text=_("Page to which this role grants permissions."),
+        on_delete=models.CASCADE,
+        limit_choices_to={
+            # permissions work with draft instances
+            "publisher_is_draft": True
+        },
+    )
+    group = models.OneToOneField(
+        to=Group,
+        related_name="role",
+        verbose_name=_("group"),
+        help_text=_("User group that this role controls."),
+        on_delete=models.PROTECT,
+        blank=True,
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "richie_page_role"
+        verbose_name = _("page role")
+        unique_together = ("page", "role")
+
+    def __str__(self):
+        """Human representation of a page role."""
+        return _("{:s} of {:s}").format(self.get_role_display(), self.page.get_title())
+
+    def create_group(self):
+        """
+        Create a user group for this role and associate it to all default permissions defined in
+        the `PAGE_ROLES` setting.
+
+        This code should only be run once for a given page role so that permissions for
+        this group can then be edited via the Django admin.
+        """
+        assert self.group_id is None
```

Comment:
Did you mean?

```python
assert self.group.id is None
```

Reply:
The `group_id` field is holding the foreign key reference so we save a useless db query.